hbase-server/src/main/java/org/apache/hadoop/hbase/mob/DefaultMobStoreCompactor.java

```diff
@@ -183,105 +270,166 @@ protected boolean performCompaction(FileDetails fd, InternalScanner scanner, Cel
-                    cellsSizeCompactedFromMob += mobCell.getCell().getValueLength();
-                  } else {
-                    // If the value of a file is empty, there might be issues when retrieving,
-                    // directly write the cell to the store file, and leave it to be handled by the
-                    // next compaction.
-                    writer.append(c);
-                  }
+                // the mob cell from the mob file, and write it back to the store file.
+                mobCell = mobStore.resolve(c, true, false);
+                if (mobCell.getCell().getValueLength() != 0) {
+                  // put the mob data back to the store file
+                  PrivateCellUtil.setSequenceId(mobCell.getCell(), c.getSequenceId());
+                  writer.append(mobCell.getCell());
+                  cellsCountCompactedFromMob++;
+                  cellsSizeCompactedFromMob += mobCell.getCell().getValueLength();
+                } else {
+                  // If the value of a file is empty, there might be issues when retrieving,
+                  // directly write the cell to the store file, and leave it to be handled by the
+                  // next compaction.
+                  LOG.error("Empty value for: " + c);
+                  writer.append(c);
+                  // Add MOB reference to a set
+                  mobRefSet.get().add(MobUtils.getMobFileName(c));
                 }
               }
             } else {
-              LOG.warn("The value format of the KeyValue " + c
-                  + " is wrong, its length is less than " + Bytes.SIZEOF_INT);
+              // TODO ????
+              LOG.error("Corrupted MOB reference: " + c);
               writer.append(c);
             }
           } else if (c.getValueLength() <= mobSizeThreshold) {
-            //If value size of a cell is not larger than the threshold, directly write to store file
+            // If the value size of a cell is not larger than the threshold, directly write it to
+            // the store file.
             writer.append(c);
           } else {
             // If the value size of a cell is larger than the threshold, it's regarded as a mob,
             // write this cell to a mob file, and write the path to the store file.
             mobCells++;
             // append the original keyValue in the mob file.
             mobFileWriter.append(c);
-            Cell reference = MobUtils.createMobRefCell(c, fileName,
-                this.mobStore.getRefCellTags());
+            Cell reference = MobUtils.createMobRefCell(c, fileName, this.mobStore.getRefCellTags());
             // write the cell whose value is the path of a mob file to the store file.
             writer.append(reference);
             cellsCountCompactedToMob++;
             cellsSizeCompactedToMob += c.getValueLength();
+            // Add ref we get for compact MOB case
+            mobRefSet.get().add(mobFileWriter.getPath().getName());
```

Comment:
we already added this?

Reply:
This is for case when compactMob = false